### PR TITLE
Implemented SADD8, UADD16, SADD16, QSASX, QSSAX, QUSAX, QUASX

### PR DESCRIPTION
**Pull request by FernandoS27**
Added and tested SADD8, UADD16, SADD16, QSASX, QSSAX, QUSAX, QUASX.

Feel free to review.

### CODE REVIEW

**Comment by FernandoS27** [comment]
as a sidenote SSUB8 and SSUB16 does not work due to an special case for the ge flags when the byte is 0x80 (-128). In ADDS8, this would be flagged as 0 since it's suppose it always generates a result under 0 (there can't be positive 128). In the case of SSUB8 doing x - (-128) would produce a circunstancial result over 0. Therefore, the current implementation is deprecated in favor to detect such case later without using microops in the IR.

File: src/backend_x64/emit_x64.cpp [hunk 1]
```diff
@@ -1277,6 +1277,21 @@ void EmitX64::EmitCountLeadingZeros(IR::Block&, IR::Inst* inst) {
...
 
+void EmitX64::EmitNegateLowWord(IR::Block&, IR::Inst* inst) {
+    IR::Value a = inst->GetArg(0);
+
+    Xbyak::Reg32 source = reg_alloc.UseScratchGpr(a).cvt32();
+    Xbyak::Reg32 result = reg_alloc.DefGpr(inst).cvt32();
+
+    code->mov(result,source);
+    code->and(result,0xFFFF0000); //save high word
+    code->sal(source,16);
```
Comment (Subv) [inline src/backend_x64/emit_x64.cpp]: Can't these 4 instructions be replaced with
````cpp
code->neg(source);
code->and(source, 0x0000FFFF);
````
?

**Comment by MerryMage** [comment]
Indents are 4 spaces in this codebase.

File: src/backend_x64/emit_x64.cpp [hunk 2]
```diff
@@ -1277,6 +1277,21 @@ void EmitX64::EmitCountLeadingZeros(IR::Block&, IR::Inst* inst) {
     }
 }
 
+void EmitX64::EmitNegateLowWord(IR::Block&, IR::Inst* inst) {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `NegateLowWord` can be implemented in four instructions:

    Xbyak::Reg32 result = reg_alloc.UseDefGpr(a, inst).cvt32();

    code->ror(result, 16);
    code->xor(result, 0xFFFF0000);
    code->add(result, 0x00010000);
    code->ror(result, 16);

File: src/backend_x64/emit_x64.cpp [hunk 3]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+    tmp = reg_alloc.ScratchGpr().cvt32();
+    code->mov(tmp,0xFFFFFFFF);
+    code->movd(xmm_b, reg_b);
+    code->movd(xmm_a, tmp);
+    //manual negate
+    if (cpu_info.has(Xbyak::util::Cpu::tSSSE3)) {
+        code->psignb(xmm_b,xmm_a);
+    }
+    else {
+        code->pxor(xmm_b, xmm_a);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Is this correct?

File: src/backend_x64/emit_x64.cpp [hunk 4]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+    if (ge_inst) {
+        EraseInstruction(block, ge_inst);
+        inst->DecrementRemainingUses();
+
+        reg_ge = reg_alloc.DefGpr(ge_inst).cvt32();
+
+        code->mov(reg_ge, reg_a);
+        code->and_(reg_ge, reg_b);
+    }
+    // SWAR Arithmetics
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: This algorithm is correct, though I do not understand the rationale behind replacing `psubb` with this.

File: src/backend_x64/emit_x64.cpp [hunk 5]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+    code->xor_(result, reg_a);
+    code->and_(result, 0x80808080);
+    code->or_(reg_a, 0x80808080);
+    code->and_(reg_b, 0x7F7F7F7F);
+    code->sub(reg_a, reg_b);
+    if (ge_inst) {
+        tmp = reg_alloc.ScratchGpr().cvt32();
+        code->mov(tmp, result);
+        code->and_(tmp, reg_a);
+        code->or_(reg_ge, tmp);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: This implementation is incorrect.

File: src/backend_x64/emit_x64.cpp [hunk 6]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+
+    code->movd(xmm_a, reg_a);
+    code->movd(xmm_b, reg_b);
+    if (ge_inst) {
+      tmp = reg_alloc.ScratchGpr().cvt32();
+      code->mov(reg_ge,reg_a);
+      code->mov(tmp,reg_a);
+      code->xor_(tmp,reg_b);
+      code->or_(reg_ge,reg_b);
+      code->not_(tmp);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: This is overly complex, and incorrect. A more obvious algorithm would to use 2 `pcmpgtb` and 1 `por`.

File: src/backend_x64/emit_x64.cpp [hunk 7]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+    code->and_(reg_a, 0x7FFF7FFF);
+    code->and_(reg_b, 0x7FFF7FFF);
+    code->add(reg_a, reg_b);
+    if (ge_inst) {
+        code->mov(tmp, result);
+        code->and_(tmp, reg_a);
+        code->or_(reg_ge, tmp);
+    }
+    code->xor_(result, reg_a);
+    if (ge_inst) {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Overly complex. Compare:

        if (ge_inst) {
            code->and_(reg_ge, 0x80008000);
            code->shr(reg_ge, 1)
            code->imul(reg_ge, reg_ge, 0xC003);
            code->shr(reg_ge, 28);
        }

File: src/backend_x64/emit_x64.cpp [hunk 8]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+
+    if (ge_inst) {
+        EraseInstruction(block, ge_inst);
+        inst->DecrementRemainingUses();
+        reg_ge = reg_alloc.DefGpr(ge_inst).cvt32();
+    }
+
+    code->movd(xmm_a, reg_a);
+    code->movd(xmm_b, reg_b);
+    if (ge_inst) {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Consider replacing with `paddsw`.

File: src/backend_x64/emit_x64.cpp [hunk 9]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+      tmp = reg_alloc.ScratchGpr().cvt32();
+      code->mov(reg_ge,reg_a);
+      code->mov(tmp,reg_a);
+      code->xor_(tmp,reg_b);
+      code->or_(reg_ge,reg_b);
+      code->not_(tmp);
+    }
+    code->paddw(xmm_a, xmm_b);
+    code->movd(reg_a, xmm_a);
+    if (ge_inst) {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Consider reducing the number of instructions to 4.

File: src/backend_x64/emit_x64.cpp [hunk 10]
```diff
@@ -1400,26 +1666,90 @@ void EmitX64::EmitPackedSubU8(IR::Block& block, IR::Inst* inst) {
...
+        code->pmaxuw(xmm_ge, xmm_b);
+        code->pcmpeqw(xmm_ge, xmm_a);
         code->movd(reg_ge, xmm_ge);
     }
-    code->psubb(xmm_a, xmm_b);
+    code->psubw(xmm_a, xmm_b);
     code->movd(reg_a, xmm_a);
 
     if (ge_inst) {
+        Xbyak::Reg32 tmp = reg_alloc.ScratchGpr().cvt32();
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: No.

File: src/backend_x64/emit_x64.cpp [hunk 11]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+    }
+    tmp = reg_alloc.ScratchGpr().cvt32();
+    code->mov(tmp,0xFFFFFFFF);
+    code->movd(xmm_b, reg_b);
+    code->movd(xmm_a, tmp);
+    //manual negate
+    if (cpu_info.has(Xbyak::util::Cpu::tSSSE3)) {
+        code->psignb(xmm_b,xmm_a);
+    }
+    else {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Style nit: `} else {`

File: src/backend_x64/emit_x64.cpp [hunk 12]
```diff
@@ -1374,12 +1389,263 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+        code->pxor(xmm_b, xmm_a);
+        code->paddb(xmm_b, xmm_a);
+    }
+    code->movd(xmm_a, reg_a);
+    if (ge_inst) {
+      code->mov(reg_ge,reg_b);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: This can be done in fewer instructions.

**Comment by MerryMage** [comment]
Please consider formatting your code to match the style of the rest of the codebase.

File: src/backend_x64/emit_x64.cpp [hunk 13]
```diff
@@ -1671,6 +2001,57 @@ void EmitX64::EmitPackedSaturatedSubS16(IR::Block&, IR::Inst* inst) {
...
+}
+
+void EmitX64::EmitPackedNegate8(IR::Block&, IR::Inst* inst) {
+    IR::Value a = inst->GetArg(0);
+
+    Xbyak::Reg32 result = reg_alloc.UseDefGpr(a, inst).cvt32();
+    Xbyak::Reg32 tmp = reg_alloc.ScratchGpr().cvt32();
+
+    Xbyak::Xmm xmm_a = reg_alloc.ScratchXmm();
+    Xbyak::Xmm xmm_b = reg_alloc.ScratchXmm();
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Rename `xmm_b`. Using this name here is confusing, there are not two parameters, a and b.

File: src/backend_x64/emit_x64.cpp [hunk 14]
```diff
@@ -1671,6 +2001,57 @@ void EmitX64::EmitPackedSaturatedSubS16(IR::Block&, IR::Inst* inst) {
...
+
+    if (cpu_info.has(Xbyak::util::Cpu::tSSSE3)) {
+        code->movd(xmm_a, result);
+        code->mov(tmp,0xFFFFFFFF);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: This is not a performant way to set an XMM register to all ones. The following is preferred:

    code->pcmpeqd(xmm_b, xmm_b);

This form is special-cased in the processor.

Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Ditto for all other instances of this.

File: src/backend_x64/emit_x64.cpp [hunk 15]
```diff
@@ -1671,6 +2001,57 @@ void EmitX64::EmitPackedSaturatedSubS16(IR::Block&, IR::Inst* inst) {
...
+        code->movd(xmm_b, tmp);
+        code->psignb(xmm_a,xmm_b);
+        code->movd(result,xmm_a);
+        return;
+    }
+    code->mov(tmp,0xFFFFFFFF);
+    code->movd(xmm_a, result);
+    code->movd(xmm_b, tmp);
+    code->pxor(xmm_a,xmm_b);
+    code->paddb(xmm_a,xmm_b);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Ditto for all other instances of this.

File: src/backend_x64/emit_x64.cpp [hunk 16]
```diff
@@ -1400,26 +1666,90 @@ void EmitX64::EmitPackedSubU8(IR::Block& block, IR::Inst* inst) {
...
+    if (ge_inst) {
+        EraseInstruction(block, ge_inst);
+        inst->DecrementRemainingUses();
+        reg_ge = reg_alloc.DefGpr(ge_inst).cvt32();
+    }
+    tmp = reg_alloc.ScratchGpr().cvt32();
+    code->mov(tmp,0xFFFFFFFF);
+    code->movd(xmm_b, reg_b);
+    code->movd(xmm_a, tmp);
+    //manual negate
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Consider not doing this.

**Comment by MerryMage** [comment]
General tips:

* Try to follow the style guide of the project you're contributing to.
* The x86_64 processor has more instructions available than currently used by the recompiler. You do not have to constrain yourself to those that are used at the moment, the only real constraint is SSE3.
* Understand code instead of copying it blindly. More optimal solutions can be found that way. Ask if you do not understand a bit twiddling hack.
* I have tried to document more tricky bits of the code especially when it may be unintuitive. Doing the same would be helpful.

File: src/backend_x64/emit_x64.cpp (hunk 16 above)
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: To clarify: A negation is unnecessary to implement this instruction.

File: src/frontend/translate/translate_arm/parallel.cpp [hunk 17]
```diff
@@ -11,33 +11,73 @@ namespace Arm {
...
 }
 
 bool ArmTranslatorVisitor::arm_SASX(Cond cond, Reg n, Reg d, Reg m) {
-    UNUSED(cond, n, d, m);
-    return InterpretThisInstruction();
+  if (d == Reg::PC || n == Reg::PC || m == Reg::PC)
+      return UnpredictableInstruction();
+  if (ConditionPassed(cond)) {
+      auto fixate = ir.RotateRight(ir.GetRegister(m),ir.Imm8(16),ir.Imm1(0));
+      auto tmp = ir.NegateLowWord(fixate.result);
```
Comment (MerryMage) [inline src/frontend/translate/translate_arm/parallel.cpp]: Using a `NegateHighWord` IR instruction would be more performant here.

File: src/frontend/translate/translate_arm/parallel.cpp [hunk 18]
```diff
@@ -11,33 +11,73 @@ namespace Arm {
...
+  return true;
 }
 
 bool ArmTranslatorVisitor::arm_SSAX(Cond cond, Reg n, Reg d, Reg m) {
-    UNUSED(cond, n, d, m);
-    return InterpretThisInstruction();
+  if (d == Reg::PC || n == Reg::PC || m == Reg::PC)
+      return UnpredictableInstruction();
+  if (ConditionPassed(cond)) {
+      auto tmp = ir.NegateLowWord(ir.GetRegister(m));
```
Comment (MerryMage) [inline src/frontend/translate/translate_arm/parallel.cpp]: Using a `NegateHighWord` IR instruction would be more performant here.

File: src/frontend/translate/translate_arm/translate_arm.h [hunk 19]
```diff
@@ -214,15 +214,8 @@ struct ArmTranslatorVisitor final {
...
-        return InterpretThisInstruction();
-    }
-    bool arm_USADA8(Cond cond, Reg d, Reg a, Reg m, Reg n) {
-        UNUSED(cond, d, a, m, n);
-        return InterpretThisInstruction();
-    }
-
+    bool arm_USAD8(Cond cond, Reg d, Reg m, Reg n);
+    bool arm_USADA8(Cond cond, Reg d, Reg a, Reg m, Reg n);
     // Packing instructions
```
Comment (MerryMage) [inline src/frontend/translate/translate_arm/translate_arm.h]: Newline.

File: tests/arm/fuzz_arm.cpp [hunk 20]
```diff
@@ -897,6 +897,31 @@ TEST_CASE("Fuzz ARM parallel instructions", "[JitX64]") {
             return halving_instructions[RandInt<size_t>(0, halving_instructions.size() - 1)].Generate();
         });
     }
+
```
Comment (MerryMage) [inline tests/arm/fuzz_arm.cpp]: Unnecessary newlines.

File: src/backend_x64/emit_x64.cpp (hunk 9 above)
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: To clarify: An `imul` with an appropriate multiplfier.

File: src/backend_x64/emit_x64.cpp (hunk 12 above)
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: To clarify: Use a saturated instruction.

File: src/backend_x64/emit_x64.cpp (hunk 15 above)
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: No. This calculates `(~a + 0xFFFF) & 0xFFFF` for each packed byte, which is not the two's complement. The two's complement would be `(~a + 1) & 0xFFFF`. 

You could chose to implement this as `(~a - 0xFFFF) & 0xFFFF` in order to reuse the constant. i.e.: replace `paddb` with `psubb`.

File: src/backend_x64/emit_x64.cpp (hunk 7 above)
Comment (FernandoS27) [inline src/backend_x64/emit_x64.cpp]: could you explain this?

**Comment by FernandoS27** [comment]
Thanks for the review and comments, I learned quite a lot.

I've done the cleanups and optimizations.

File: src/frontend/translate/translate_arm/translate_arm.h (hunk 19 above)
Comment (MerryMage) [inline src/frontend/translate/translate_arm/translate_arm.h]: Newline before `//Packing instructions`

File: tests/arm/fuzz_arm.cpp [hunk 21]
```diff
@@ -897,6 +915,31 @@ TEST_CASE("Fuzz ARM parallel instructions", "[JitX64]") {
...
+
+}
+
+TEST_CASE("Fuzz ARM sum of absolute differences", "[JitX64]") {
+  auto validate_d_m_n = [](u32 inst) -> bool {
```
Comment (MerryMage) [inline tests/arm/fuzz_arm.cpp]: Indent to 4 spaces.

File: tests/arm/fuzz_arm.cpp [hunk 22]
```diff
@@ -897,6 +915,31 @@ TEST_CASE("Fuzz ARM parallel instructions", "[JitX64]") {
             return halving_instructions[RandInt<size_t>(0, halving_instructions.size() - 1)].Generate();
         });
     }
+
+
+}
```
Comment (MerryMage) [inline tests/arm/fuzz_arm.cpp]: Remove the unnecessary newlines before the `}`.

File: src/backend_x64/emit_x64.cpp [hunk 23]
```diff
@@ -1277,6 +1277,26 @@ void EmitX64::EmitCountLeadingZeros(IR::Block&, IR::Inst* inst) {
     }
 }
 
+void EmitX64::EmitNegateLowWord(IR::Block&, IR::Inst* inst) {
+    IR::Value a = inst->GetArg(0);
+
+    Xbyak::Reg32 result = reg_alloc.UseDefGpr(a,inst).cvt32();
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `a,inst` → `a, inst`

File: src/backend_x64/emit_x64.cpp [hunk 24]
```diff
@@ -1277,6 +1277,26 @@ void EmitX64::EmitCountLeadingZeros(IR::Block&, IR::Inst* inst) {
...
+    code->ror(result, 16);
+    code->xor(result, 0xFFFF0000);
+    code->add(result, 0x00010000);
+    code->ror(result, 16);
+}
+
+void EmitX64::EmitNegateHighWord(IR::Block&, IR::Inst* inst) {
+    IR::Value a = inst->GetArg(0);
+
+    Xbyak::Reg32 result = reg_alloc.UseDefGpr(a,inst).cvt32();
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `a,inst` → `a, inst`

File: src/backend_x64/emit_x64.cpp [hunk 25]
```diff
@@ -1408,15 +1552,115 @@ void EmitX64::EmitPackedSubU8(IR::Block& block, IR::Inst* inst) {
...
+
+void EmitX64::EmitPackedSubS8(IR::Block& block, IR::Inst* inst) {
+    auto ge_inst = inst->GetAssociatedPseudoOperation(IR::Opcode::GetGEFromOp);
+
+    IR::Value a = inst->GetArg(0);
+    IR::Value b = inst->GetArg(1);
+
+    Xbyak::Reg32 reg_a = reg_alloc.UseDefGpr(a, inst).cvt32();
+    Xbyak::Reg32 reg_b = reg_alloc.UseGpr(b).cvt32();
+    Xbyak::Reg32 reg_ge,tmp;
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `Xbyak::Reg32 reg_ge,tmp;` → `Xbyak::Reg32 reg_ge;`

`tmp` appears to be unused is this function.

File: src/backend_x64/emit_x64.cpp [hunk 26]
```diff
@@ -1408,15 +1552,115 @@ void EmitX64::EmitPackedSubU8(IR::Block& block, IR::Inst* inst) {
...
+    if (ge_inst) {
+        code->movdqa(xmm_ge, xmm_a);
+        code->pmaxuw(xmm_ge, xmm_b);
+        code->pcmpeqw(xmm_ge, xmm_a);
+        code->movd(reg_ge, xmm_ge);
+    }
+    code->psubw(xmm_a, xmm_b);
+    code->movd(reg_a, xmm_a);
+    if (ge_inst) {
+        ExtractMostSignificantBitFromPackedWords(code,reg_ge);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `code,reg_ge` → `code, reg_ge`

File: src/backend_x64/emit_x64.cpp [hunk 27]
```diff
@@ -1408,15 +1552,115 @@ void EmitX64::EmitPackedSubU8(IR::Block& block, IR::Inst* inst) {
...
+        Xbyak::Xmm xmm_ge = reg_alloc.ScratchXmm();
+        code->movdqa(xmm_ge, xmm_a);
+        code->psubsw(xmm_ge, xmm_b);
+        code->movd(reg_ge, xmm_ge);
+    }
+    code->psubw(xmm_a, xmm_b);
+    code->movd(reg_a, xmm_a);
+    if (ge_inst) {
+        code->not_(reg_ge);
+        ExtractMostSignificantBitFromPackedWords(code,reg_ge);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `code,reg_ge` → `code, reg_ge`

File: src/frontend/ir/opcodes.inc [hunk 28]
```diff
@@ -73,14 +73,22 @@ OPCODE(ByteReverseWord,         T::U32,         T::U32
...
 OPCODE(CountLeadingZeros,       T::U32,         T::U32                                          )
+OPCODE(NegateLowWord,           T::U32,         T::U32                                          )
+OPCODE(NegateHighWord,          T::U32,         T::U32                                          )
 
 // Saturated instructions
 OPCODE(SignedSaturatedAdd,      T::U32,         T::U32,         T::U32                          )
 OPCODE(SignedSaturatedSub,      T::U32,         T::U32,         T::U32                          )
 
 // Packed instructions
+OPCODE(PackedAddS8,             T::U32,         T::U32,         T::U32                          )
```
Comment (MerryMage) [inline src/frontend/ir/opcodes.inc]: Ordering: U8 before S8, U16 before S16.

    OPCODE(PackedAddU8,             T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedAddS8,             T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedSubU8,             T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedSubS8,             T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedAddU16,            T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedAddS16,            T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedSubU16,            T::U32,         T::U32,         T::U32                          )
    OPCODE(PackedSubS16,            T::U32,         T::U32,         T::U32                          )

File: src/backend_x64/emit_x64.cpp [hunk 29]
```diff
@@ -1363,14 +1402,119 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+    code->add(reg_a, reg_b);
+    if (ge_inst) {
+        tmp = reg_alloc.ScratchGpr().cvt32();
+        code->mov(tmp, result);
+        code->and_(tmp, reg_a);
+        code->or_(reg_ge, tmp);
+    }
+    code->xor_(result, reg_a);
+    if (ge_inst) {
+        ExtractMostSignificantBitFromPackedWords(code,reg_ge);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `code,reg_ge` → `code, reg_ge`

File: src/backend_x64/emit_x64.cpp [hunk 30]
```diff
@@ -1363,14 +1402,119 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+        Xbyak::Xmm saturated_sum = reg_alloc.ScratchXmm();
+        code->movdqa(saturated_sum, xmm_a);
+        code->paddsw(saturated_sum, xmm_b);
+        code->movd(reg_ge, saturated_sum);
+    }
+    code->paddw(xmm_a, xmm_b);
+    code->movd(reg_a, xmm_a);
+    if (ge_inst) {
+        code->not_(reg_ge);
+        ExtractMostSignificantBitFromPackedWords(code,reg_ge);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `code,reg_ge` → `code, reg_ge`

File: src/backend_x64/emit_x64.cpp [hunk 31]
```diff
@@ -1363,14 +1402,119 @@ void EmitX64::EmitPackedAddU8(IR::Block& block, IR::Inst* inst) {
...
+
+void EmitX64::EmitPackedAddS16(IR::Block& block, IR::Inst* inst) {
+    auto ge_inst = inst->GetAssociatedPseudoOperation(IR::Opcode::GetGEFromOp);
+
+    IR::Value a = inst->GetArg(0);
+    IR::Value b = inst->GetArg(1);
+
+    Xbyak::Reg32 reg_a = reg_alloc.UseDefGpr(a, inst).cvt32();
+    Xbyak::Reg32 reg_b = reg_alloc.UseGpr(b).cvt32();
+    Xbyak::Reg32 reg_ge, tmp;
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `tmp` appears to be unused in this function. Remove if it is unused.

File: src/backend_x64/emit_x64.cpp [hunk 32]
```diff
@@ -1408,15 +1552,115 @@ void EmitX64::EmitPackedSubU8(IR::Block& block, IR::Inst* inst) {
...
+
+void EmitX64::EmitPackedSubS16(IR::Block& block, IR::Inst* inst) {
+    auto ge_inst = inst->GetAssociatedPseudoOperation(IR::Opcode::GetGEFromOp);
+
+    IR::Value a = inst->GetArg(0);
+    IR::Value b = inst->GetArg(1);
+
+    Xbyak::Reg32 reg_a = reg_alloc.UseDefGpr(a, inst).cvt32();
+    Xbyak::Reg32 reg_b = reg_alloc.UseGpr(b).cvt32();
+    Xbyak::Reg32 reg_ge,tmp;
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: `tmp` appears to be unused in this function.

**Comment by lioncash** [comment]
It would also be nice if 

- merge commits weren't included in your PR list (look up rebasing for git)
- All of your commits had actual informative names.

Consider someone looking through git's logs for changes. When they come across a commit named `"First batch of fixes based on reviews"`, `"More cleaning and refactor"` or `"Ordering."` - they aren't very helpful nor descriptive about what exactly is being changed.

`"Fix compiling"` really shouldn't even be in this PR; rather, it should be in a completely different one, as it's not related directly to implementing any of the instructions.

**Comment by MerryMage** [comment]
Smaller PRs are easier to review than larger ones. Please do not continue to add more things to the same PR.

**Comment by FernandoS27** [comment]
Thank you, I'll try to make future PRs a lot smaller

File: src/frontend/translate/translate_arm/parallel.cpp [hunk 33]
```diff
@@ -52,18 +106,53 @@ bool ArmTranslatorVisitor::arm_UADD8(Cond cond, Reg n, Reg d, Reg m) {
...
+    return true;
 }
 
 bool ArmTranslatorVisitor::arm_USAX(Cond cond, Reg n, Reg d, Reg m) {
-    UNUSED(cond, n, d, m);
-    return InterpretThisInstruction();
+    if (d == Reg::PC || n == Reg::PC || m == Reg::PC)
+        return UnpredictableInstruction();
+    if (ConditionPassed(cond)) {
+        GESubstractAddExchange(n, d, m, &ir, &IR::IREmitter::PackedAddU16);
```
Comment (MerryMage) [inline src/frontend/translate/translate_arm/parallel.cpp]: This implementation fails when `Rm<31:16> == 0x0000`.

Comment (FernandoS27) [inline src/frontend/translate/translate_arm/parallel.cpp]: I think deactivated those two on last commit, I'll rebase and fix those two on a sepparate branch.

Comment (MerryMage) [inline src/frontend/translate/translate_arm/parallel.cpp]: For an example, see 975987f38e8f51d8a02b1fcc8b43928d79622495.

File: src/frontend/translate/translate_arm/parallel.cpp [hunk 34]
```diff
@@ -203,13 +312,21 @@ bool ArmTranslatorVisitor::arm_SHADD16(Cond cond, Reg n, Reg d, Reg m) {
...
+    return true;
 }
 
 bool ArmTranslatorVisitor::arm_SHSAX(Cond cond, Reg n, Reg d, Reg m) {
-    UNUSED(cond, n, d, m);
-    return InterpretThisInstruction();
+    if (d == Reg::PC || n == Reg::PC || m == Reg::PC)
+        return UnpredictableInstruction();
+    if (ConditionPassed(cond)) {
+        SubstractAddExchange(n, d, m, &ir, &IR::IREmitter::PackedHalvingAddS16);
```
Comment (MerryMage) [inline src/frontend/translate/translate_arm/parallel.cpp]: This implementation is incorrect. For a failing testcase, see 821b26e9d21db8817a7b457f6893663f57b32828.

Comment (FernandoS27) [inline src/frontend/translate/translate_arm/parallel.cpp]: Thanks, I'll deactivate them. Looks like only the saturated ones should work. When the PR be merged, I'll rebase a branch each pair. (2 instructions per PR)

**Comment by MerryMage** [comment]
Chose to remove implementations of QASX, QSAX, UQASX, UQSAX as those were untested. Feel free to resubmit them in another PR.

**Pull request merged**